engine/consensus/matching/engine.go

```diff
@@ -651,37 +703,75 @@ func (e *Engine) sealableResults() ([]*flow.IncorporatedResult, error) {
 		// matching engine must enforce equality of start and end state for a result with zero chunks,
 		// in the absence of anyone else doing do.
 		matched := false
+		unmatchedIndex := -1
 		// check that each chunk collects enough approvals
-		for _, chunk := range incorporatedResult.Result.Chunks {
+		for i, chunk := range incorporatedResult.Result.Chunks {
 			matched, err = e.matchChunk(incorporatedResult, block, chunk, assignment)
 			if err != nil {
-				return nil, fmt.Errorf("could not match chunk: %w", err)
+				return nil, nil, fmt.Errorf("could not match chunk: %w", err)
 			}
 			if !matched {
+				unmatchedIndex = i
 				break
 			}
 		}
 
+		matchedAfterConsideringEmergency := matched
 		// ATTENTION: this is a temporary solution called emergency sealing. Emergency sealing is a special case
 		// when we seal ERs that don't have enough approvals but are deep enough in the chain resulting in halting sealing
 		// process. This will be removed when implementation of seal & verification is finished.
 		if !matched && e.emergencySealingActive {
-			block, err := e.headersDB.ByBlockID(incorporatedResult.IncorporatedBlockID)
+			incorporatedBlock, err := e.headersDB.ByBlockID(incorporatedResult.IncorporatedBlockID)
 			if err != nil {
-				return nil, fmt.Errorf("could not get block %v: %w", incorporatedResult.IncorporatedBlockID, err)
+				return nil, nil, fmt.Errorf("could not get block %v: %w", incorporatedResult.IncorporatedBlockID, err)
 			}
-			if block.Height+DefaultEmergencySealingThreshold <= lastFinalized.Height {
-				matched = true
+			// Note:
+			// we assume the incorporatedBlock is for a unsealed block, if
+			// there are DefaultEmergencySealingThreshold number of blocks between incorporatedBlock
+			// and lastFinalized, it means there are as many unsealed and finalized blocks, because
+			// should trigger the emergency sealing
+			if incorporatedBlock.Height+DefaultEmergencySealingThreshold <= lastFinalized.Height {
+				matchedAfterConsideringEmergency = true
 			}
 		}
 
-		if matched {
+		if matchedAfterConsideringEmergency {
 			// add the result to the results that should be sealed
 			results = append(results, incorporatedResult)
 		}
+
+		// whether we sealed the result when the emergency sealing is active and triggered
+		sealedByEmergency := !matched && matchedAfterConsideringEmergency
+
+		if nextUnsealedIsFinalized {
+			if incorporatedResult.Result.BlockID == nextUnsealed {
+				nextUnsealeds = append(nextUnsealeds, &nextUnsealedResult{
+					BlockID:                  incorporatedResult.Result.BlockID,
+					Height:                   block.Height,
+					ResultID:                 incorporatedResult.Result.ID(),
+					IncorporatedResultID:     incorporatedResult.ID(),
```

Comment:
The incorporated result only exists in the mempool. Therefore, its `incorporatedResult.ID()` is of little value as, we can't query it from the database. However, the block which _incorporates_ the result is much more valuable to know:
```suggestion
					IncorporatedBlockID:     incorporatedResult.IncorporatedBlockID,
```

Reply:
The IncorporatedResult.ID() is used [here by the builder to log if a seal](https://github.com/onflow/flow-go/pull/363/files#diff-17e93dc346462f66b95ec3f31134583765b8fdc4eff5e3c03f276698a554997fR621) is indeed added to the mempool, and available for the builder to include.

I can't use the `incorporatedResult.IncorporatedBlockID`, since the incorporatedResultSeal's ID includes both it and the result id. And I don't want to duplicate that `ID` logic in the logging.